src/Bicep.Core/Emit/InlineDependencyVisitor.cs

```diff
@@ -198,29 +198,44 @@ private void VisitPropertyAccessSyntaxInternal(PropertyAccessSyntax syntax)
                 return;
             }
 
-            static bool ShouldSkipInlining(ObjectType objectType, string propertyName)
+            static bool ShouldSkipInlining(ObjectType objectType, string propertyName, ResourceDeclarationSyntax? resourceSyntax = null)
             {
                 if (!objectType.Properties.TryGetValue(propertyName, out var propertyType))
                 {
                     // unknown property - we should assume it's not inlinable
                     return true;
                 }
 
-                // update the cache if property can't be skipped for inlining
-                return propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant);
+                if (propertyType.Flags.HasFlag(TypePropertyFlags.DeployTimeConstant))
+                {
+                    if (resourceSyntax is not null &&
+                        propertyName != LanguageConstants.ResourceIdPropertyName &&
+                        propertyName != LanguageConstants.ResourceNamePropertyName &&
+                        propertyName != LanguageConstants.ResourceTypePropertyName &&
+                        propertyName != LanguageConstants.ResourceApiVersionPropertyName &&
+                        resourceSyntax.TryGetBody()?.SafeGetPropertyByName(propertyName) is null)
+                    {
+                        // The property is not declared in the resource - we should inline it even it is a deploy-time constant.
+                        return false;
+                    }
+
+                    return true;
+                }
+
+                return false;
             }
 
             switch (model.GetSymbolInfo(variableAccessSyntax))
             {
                 // Note - there's a limitation here that we're using the 'declared' type and not the 'assigned' type.
                 // This means that we may encounter a DiscriminatedObjectType. For now we should accept this limitation,
                 // and move to using the assigned type once https://github.com/Azure/bicep/issues/1177 is fixed.
```

Comment:
I think the changes under #1915 **may** have fixed this, but I'll verify that separately - no need to change this PR.